File: app/src/main/java/com/cybersiren/android/v2v/auto/V2VCarServiceHolder.kt
package com.cybersiren.android.v2v.auto

import com.cybersiren.android.v2v.model.AlertMode
import com.cybersiren.android.v2v.model.ReceivedAlert
import com.cybersiren.android.v2v.model.VehicleType

interface V2VCarServiceInterface {
    fun getActiveAlerts(): List<ReceivedAlert>
    fun getAlertMode(): AlertMode
    fun setMode(mode: AlertMode)
    fun isEmergencyActive(): Boolean
    fun toggleEmergencyBroadcast()
    fun startEmergencyBroadcast(vehicleType: VehicleType)
    fun stopEmergencyBroadcast()
    fun getConnectedPeers(): Int

    fun getSelectedVehicleType(): VehicleType
    fun selectVehicleType(type: VehicleType)
    fun getCurrentLatitude(): Double?
    fun getCurrentLongitude(): Double?
    fun getCurrentSpeedKmh(): Float
    fun getCurrentHeadingDegrees(): Float
}

object V2VCarServiceHolder {
    @Volatile
    private var service: V2VCarServiceInterface? = null

    fun setService(service: V2VCarServiceInterface?) {
        this.service = service
    }

    fun getService(): V2VCarServiceInterface? = service

    fun isServiceAvailable(): Boolean = service != null
}

File: ios/bitchat/bitchat/V2V/CarPlay/V2VCarServiceHolder.swift
import Foundation
import CoreLocation
import Combine

public protocol V2VCarService: AnyObject {
    func getActiveAlerts() -> [ReceivedAlert]
    func getAlertMode() -> AlertMode
    func isEmergencyActive() -> Bool
    func getConnectedPeers() -> Int
    func getSelectedVehicleType() -> VehicleType
    func getCurrentLatitude() -> Double?
    func getCurrentLongitude() -> Double?
    func getCurrentSpeedKmh() -> Float
    func getCurrentHeadingDegrees() -> Float

    func setMode(_ mode: AlertMode)
    func selectVehicleType(_ type: VehicleType)
    func startEmergencyBroadcast(vehicleType: VehicleType)
    func stopEmergencyBroadcast()
    func toggleEmergencyBroadcast()
}

public final class V2VCarServiceHolder {
    public static let shared = V2VCarServiceHolder()
    private let lock = NSLock()
    private weak var service: V2VCarService?
    private init() {}

    public func setService(_ service: V2VCarService?) {
        lock.lock(); defer { lock.unlock() }
        self.service = service
        NotificationCenter.default.post(name: .v2vCarServiceChanged, object: nil)
    }

    public func getService() -> V2VCarService? {
        lock.lock(); defer { lock.unlock() }
        return service
    }
}

public extension Notification.Name {
    static let v2vCarServiceChanged = Notification.Name("v2v.carServiceChanged")
}
